### PR TITLE
refactored a little bit of the code

### PR DESCRIPTION
**Pull request by dmitriy-p-petrov**
Hi! I looked at another ticket. I am leaving this here just so you can check it out and tell me if you don't like something while i'm still in the very beginning. 

What's done so far is:
- I added a couple more routes that do nothing for now;
- I changed `populateTemplate()` and all the helper functions it uses. I wanted to reuse them to populate the graphs for the other routes ( still not sure what to do with `makeColumnNames()` ).

And that's pretty much it. Everything else is just rearrangements.

After i figure out what to do with `makeColumnNames()` i'm gonna work on the `collector()` function. I'm just gonna add a few more checks for the other events. I'm not sure what to do with the `route-metrics.log` path. Maybe just receive it as a command-line argument?

### CODE REVIEW

File: servers/express.js [hunk 1]
```diff
@@ -2,32 +2,39 @@
...
 
 const Skeleton = require('./skeleton');
 const watcher = require('../file-watcher.js');
 
 // the app
-const express = require('express');
 const app = express();
 
-const pathToLogFile = path.join(__dirname, '..', '..', 'node-route-metrics', 'route-metrics.log');
+const pathToLogFile = (process.argv.length > 2) ? process.argv[2] : 'router-metrics.log';
```
Comment (bmacnaughton) [inline servers/express.js:13]: route-metrics, not router-metrics

File: servers/express.js [hunk 2]
```diff
@@ -2,32 +2,39 @@
...
 const htmlTemplate = fs.readFileSync(path.join(__dirname, 'pages/templatized.html'), 'utf8');
 
 app.get('/eventloop', function(req, res) {
   const chartOpt = makeChartOptions({title: 'Eventloop lag percentiles (ms)', subtitle: '@contrast/route-metrics'});
   const datarows = makeDataRows(eventloopDataRows);
-  const columns  = makeColumnNames();
-
-  const html = populateTemplate(htmlTemplate, chartOpt, columns, datarows);
+  
+  const html = populateTemplate(htmlTemplate, chartOpt, makeEventloopColumnNames(), datarows);
```
Comment (bmacnaughton) [inline servers/express.js:20]: probably want to make event loop column names the same form as data rows, e.g.,

`const columnNames = makeEventloopColumnNames();` and then use that in the call to `populateTemplate()`.

File: servers/express.js [hunk 3]
```diff
@@ -2,32 +2,39 @@
...
   res.send(html);
 });
 
-app.get('/processor', function(req, res) {
-  res.send('Work in progress');
+app.get('/memory', function(req, res) {
+  const chartOpt = makeChartOptions({title: 'Insert title here', subtitle: '@contrast/route-metrics'});
+  const datarows = makeDataRows(memoryDataRows);
+  
+  const html = populateTemplate(htmlTemplate, chartOpt, makeMemoryColumnNames(), datarows);
```
Comment (bmacnaughton) [inline servers/express.js:28]: ditto

File: servers/express.js [hunk 4]
```diff
@@ -2,32 +2,39 @@
...
+  res.send(html);
 });
 
-app.get('/memory', function(req, res) {
-  res.send('Work in progress');
+app.get('/cpu', function(req, res) {
+  const chartOpt = makeChartOptions({title: 'Time spent in user and system code respectively (ms)', subtitle: '@contrast/route-metrics'});
```
Comment (bmacnaughton) [inline servers/express.js:33]: it's always good to have the units in the chart - nice.

File: servers/express.js [hunk 5]
```diff
@@ -65,51 +84,51 @@ const eventloopDataRows = [];
...
+        // Memory data is in bytes. Make it megabytes
+        const externalAvg = entry['externalAvg'] / 1e6;
+        const heapUsedAvg = entry['heapUsedAvg'] / 1e6;
+        const heapTotal = entry['heapTotal'] / 1e6;
+        const rss = entry['rss'] / 1e6;
+
+        // Cpu data is in microseconds. Make it ms
+        cpuDataRows.push(`[${delta}, ${entry['cpuUserAvg'] / 1e3}, ${entry['cpuSystemAvg'] / 1e3}]`);
+        memoryDataRows.push(`[${delta}, ${rss}, ${heapTotal}, ${heapUsedAvg}, ${externalAvg}]`);
+      } else continue;
```
Comment (bmacnaughton) [inline servers/express.js:121]: this i definitely don't like - to me `else` should always follow the form of the previous if/else-if blocks for consistency, so

```` 
} else {
     continue;
}
```

**Review by bmacnaughton — CHANGES_REQUESTED** [review]

**Comment by dmitriy-p-petrov** [comment]
Thanks for the feedback! I addressed all of your points.

About the `undefined` thing - i know it's `undefined` by default and it doesn't really matter, since it's used on the next line. I just like to explicitly define them even when they are empty.

I believe the arguments problem is fixed now. You can pass `-l` (it's an L) or `--logfile` for the logfile and `-p` or `--protocol` for the protocols. You can either pass one, two or neither of them. (i mean, you do kinda need to pass at least -p currently, because both servers use the same port by default, but yeah...)

I'm gonna finish the last chart now. I wanted to ask if all of them represent the same thing on the x-axis and what it is exactly? Looks like they are functions with respect to the number of entries in the logfile, but it said `seconds` down there, so i got kinda confused.
I'm also not sure if i should leave the `makeColumnNames()` functions so static?

**Comment by bmacnaughton** [comment]
> I'm also not sure if i should leave the makeColumnNames() functions so static?

i don't understand this question.

>I'm gonna finish the last chart now. I wanted to ask if all of them represent the same thing on the x-axis and what it is exactly? Looks like they are functions with respect to the number of entries in the logfile, but it says seconds down there, so i got kinda confused.

yes, the x-axis is always time. that's why we're subtracting the base timestamp from each log entry's timestamp (then divide by 1000, because the times are ms). it might be nice to note in the x-axis title the base time, so something like:

`seconds since 2022-06-09T15:05:11.277Z`

**Comment by bmacnaughton** [comment]
> I believe the arguments problem is fixed now. You can pass -l (it's an L) or --logfile for the logfile and -p or --protocol for the protocols. You can either pass one, two or neither of them. (i mean, you do kinda need to pass at least -p currently, because both servers use the same port by default, but yeah...)

thinking about this, it should probably take the protocol argument on the command line (not an option) because it's required. that way it will continue to work as it is but the logfile is now an optional argument.

**Comment by dmitriy-p-petrov** [comment]
Hello!

I'm once again asking for more details.
Looked at the ticket again. I think the first few points are done to some extend.

- Not sure what "configurable port" means, though... isn't it already configurable, since it's not really optional currently?
- I also don't understand timed-fetch part

**Review by bmacnaughton — APPROVED** [review]

**Pull request merged**